### ISSUE TITLE
Thread-unsafe handling of pingResponse in network.go

### ISSUE THREAD

**Issue by AvSquirrel** [cyoung/stratux#568]
1. Stratux version: 1.2r1 + commmits (0ee94ff925a2d5)
2. Stratux config:

    SDR
      - [ ] single
      - [x] dual

    GPS
      - [x] yes
      - [ ] no
      type: RY725AI over USB (equivalent to VK-162)

    AHRS
      - [ ] yes
      - [x] no

    power source: Wall-wart adapter

3. EFB app and version: ForeFlight 8.3 (sleeping on two devices)
    EFB platform: iOS 10.2.1
    EFB hardware: iPad Air 2 and iPhone 6 (both devices sleeping)

4. Description of your issue:

I was running a long term test on commit 0ee94ff925 to evaluate UAT TIS-B suppression. A little over two days into the run, gen_gdl90 crashed. The stack trace indicated that there is an unsafe handling of the `pingResponse` map in network.go

```
fatal error: concurrent map read and map write

goroutine 49 [running]:
runtime.throw(0x4d7f57, 0x21)
        /root/go/src/runtime/panic.go:566 +0x78 fp=0x10a9eb0c sp=0x10a9eb00
runtime.mapaccess2_faststr(0x47c578, 0x109e2820, 0x109f9800, 0xd, 0x13152d80, 0x2)
        /root/go/src/runtime/hashmap_fast.go:306 +0x64 fp=0x10a9eb3c sp=0x10a9eb0c
main.isSleeping(0x109f9800, 0x12, 0x10a9ec30)
        /root/stratux-dev/main/network.go:151 +0xa4 fp=0x10a9ebcc sp=0x10a9eb3c
main.sendToAllConnectedClients(0x1427f3a0, 0x14, 0x20, 0x4, 0x3d4d6, 0x0, 0x1c9c380, 0x0)
        /root/stratux-dev/main/network.go:176 +0x148 fp=0x10a9ed44 sp=0x10a9ebcc
main.messageQueueSender()
        /root/stratux-dev/main/network.go:367 +0x10c fp=0x10a9efd4 sp=0x10a9ed44
runtime.goexit()
        /root/go/src/runtime/asm_arm.s:998 +0x4 fp=0x10a9efd4 sp=0x10a9efd4
created by main.initNetwork
        /root/stratux-dev/main/network.go:644 +0x1a8
```

The network code in my test build is the same as 1.2r1, and is unchanged since 6198fc2  (12/25/2016). [`isSleeping()`](https://github.com/cyoung/stratux/blob/6198fc2ce34f8b230066d42d21672784ef19cd8d/main/network.go#L148-L158) contained the read of `pingResponse` that triggered this crash; there also appears to be an unprotected read in [`getNetworkStats()`](https://github.com/cyoung/stratux/blob/6198fc2ce34f8b230066d42d21672784ef19cd8d/main/network.go#L304) and an unprotected write in [`sleepMonitor()`](https://github.com/cyoung/stratux/blob/6198fc2ce34f8b230066d42d21672784ef19cd8d/main/network.go#L524).

```
func isSleeping(k string) bool {
	ipAndPort := strings.Split(k, ":")
	// No ping response. Assume disconnected/sleeping device.
	if lastPing, ok := pingResponse[ipAndPort[0]]; !ok || stratuxClock.Since(lastPing) > (10*time.Second) {
		return true
	}
	if stratuxClock.Since(outSockets[k].LastUnreachable) < (5 * time.Second) {
		return true
	}
	return false
}
```

Full stack trace:
```
fatal error: concurrent map read and map write

goroutine 49 [running]:
runtime.throw(0x4d7f57, 0x21)
        /root/go/src/runtime/panic.go:566 +0x78 fp=0x10a9eb0c sp=0x10a9eb00
runtime.mapaccess2_faststr(0x47c578, 0x109e2820, 0x109f9800, 0xd, 0x13152d80, 0x2)
        /root/go/src/runtime/hashmap_fast.go:306 +0x64 fp=0x10a9eb3c sp=0x10a9eb0c
main.isSleeping(0x109f9800, 0x12, 0x10a9ec30)
        /root/stratux-dev/main/network.go:151 +0xa4 fp=0x10a9ebcc sp=0x10a9eb3c
main.sendToAllConnectedClients(0x1427f3a0, 0x14, 0x20, 0x4, 0x3d4d6, 0x0, 0x1c9c380, 0x0)
        /root/stratux-dev/main/network.go:176 +0x148 fp=0x10a9ed44 sp=0x10a9ebcc
main.messageQueueSender()
        /root/stratux-dev/main/network.go:367 +0x10c fp=0x10a9efd4 sp=0x10a9ed44
runtime.goexit()
        /root/go/src/runtime/asm_arm.s:998 +0x4 fp=0x10a9efd4 sp=0x10a9efd4
created by main.initNetwork
        /root/stratux-dev/main/network.go:644 +0x1a8

goroutine 1 [select (no cases), 4186 minutes]:
main.main()
        /root/stratux-dev/main/gen_gdl90.go:1482 +0xd90

goroutine 17 [syscall, 4186 minutes, locked to thread]:
runtime.goexit()
        /root/go/src/runtime/asm_arm.s:998 +0x4

goroutine 5 [syscall, 4186 minutes]:
os/signal.signal_recv(0x0)
        /root/go/src/runtime/sigqueue.go:116 +0x190
os/signal.loop()
        /root/go/src/os/signal/signal_unix.go:22 +0x14
created by os/signal.init.1
        /root/go/src/os/signal/signal_unix.go:28 +0x30

goroutine 6 [chan receive]:
github.com/golang/glog.(*loggingT).flushDaemon(0x6698c8)
        /root/go_path/src/github.com/golang/glog/glog.go:882 +0x60
created by github.com/golang/glog.init.1
        /root/go_path/src/github.com/golang/glog/glog.go:410 +0x1ec

goroutine 7 [select, 4186 minutes, locked to thread]:
runtime.gopark(0x50ec0c, 0x0, 0x4cb84e, 0x6, 0x18, 0x2)
        /root/go/src/runtime/proc.go:259 +0x148
runtime.selectgoImpl(0x10924f90, 0x0, 0xc)
        /root/go/src/runtime/select.go:423 +0x1368
runtime.selectgo(0x10924f90)
        /root/go/src/runtime/select.go:238 +0x10
runtime.ensureSigM.func1()
        /root/go/src/runtime/signal1_unix.go:304 +0x428
runtime.goexit()
        /root/go/src/runtime/asm_arm.s:998 +0x4

goroutine 34 [chan receive, 4186 minutes]:
main.signalWatcher()
        /root/stratux-dev/main/gen_gdl90.go:1313 +0x3c
created by main.main
        /root/stratux-dev/main/gen_gdl90.go:1338 +0x120

goroutine 35 [runnable]:
main.(*monotonic).Watcher(0x109de030)
        /root/stratux-dev/main/monotonic.go:30 +0x34
created by main.NewMonotonic
        /root/stratux-dev/main/monotonic.go:64 +0x9c

goroutine 36 [sleep]:
time.Sleep(0x3b9aca00, 0x0)
        /root/go/src/runtime/time.go:59 +0xec
main.sdrWatcher()
        /root/stratux-dev/main/sdr.go:475 +0x3c
created by main.sdrInit
        /root/stratux-dev/main/sdr.go:538 +0x24

goroutine 37 [chan receive]:
main.uatReader()
        /root/stratux-dev/main/sdr.go:311 +0xbc
created by main.sdrInit
        /root/stratux-dev/main/sdr.go:539 +0x38

goroutine 38 [syscall, locked to thread]:
_/root/stratux-dev/godump978._Cfunc_process_data(0x116a6000, 0x40000, 0x0)
        _/root/stratux-dev/godump978/_obj/_cgo_gotypes.go:78 +0x3c
_/root/stratux-dev/godump978.ProcessData(0x116a6000, 0x40000, 0x40000)
        /root/stratux-dev/godump978/godump978.go:53 +0x40
_/root/stratux-dev/godump978.ProcessDataFromChannel()
        /root/stratux-dev/godump978/godump978.go:59 +0x68
created by main.sdrInit
        /root/stratux-dev/main/sdr.go:540 +0x4c

goroutine 39 [sleep]:
time.Sleep(0x3b9aca00, 0x0)
        /root/go/src/runtime/time.go:59 +0xec
main.pingWatcher()
        /root/stratux-dev/main/ping.go:233 +0x2c
created by main.pingInit
        /root/stratux-dev/main/ping.go:267 +0x24

goroutine 40 [IO wait]:
net.runtime_pollWait(0x740c4a98, 0x72, 0x10a94000)
        /root/go/src/runtime/netpoll.go:160 +0x60
net.(*pollDesc).wait(0x109971f8, 0x72, 0x0, 0x0)
        /root/go/src/net/fd_poll_runtime.go:73 +0x34
net.(*pollDesc).waitRead(0x109971f8, 0x0, 0x0)
        /root/go/src/net/fd_poll_runtime.go:78 +0x30
net.(*netFD).Read(0x109971c0, 0x10a94000, 0x1000, 0x1000, 0x0, 0x6536b0, 0x109100c0)
        /root/go/src/net/fd_unix.go:243 +0x1f0
net.(*conn).Read(0x1099c260, 0x10a94000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
        /root/go/src/net/net.go:173 +0xb8
bufio.(*Reader).fill(0x11281cd8)
        /root/go/src/bufio/bufio.go:97 +0x1c4
bufio.(*Reader).ReadSlice(0x11281cd8, 0xa, 0x0, 0x0, 0x0, 0x0, 0x0)
        /root/go/src/bufio/bufio.go:330 +0x264
bufio.(*Reader).ReadBytes(0x11281cd8, 0xa, 0x0, 0x0, 0x0, 0x0, 0x0)
        /root/go/src/bufio/bufio.go:408 +0x80
bufio.(*Reader).ReadString(0x11281cd8, 0x109e220a, 0x0, 0x0, 0x0, 0x0)
        /root/go/src/bufio/bufio.go:448 +0x3c
main.esListen()
        /root/stratux-dev/main/traffic.go:731 +0x1b0
created by main.initTraffic
        /root/stratux-dev/main/traffic.go:1294 +0xe8

goroutine 41 [sleep]:
time.Sleep(0x3b9aca00, 0x0)
        /root/go/src/runtime/time.go:59 +0xec
main.dataLogWatchdog()
        /root/stratux-dev/main/datalog.go:611 +0x7c
created by main.initDataLog
        /root/stratux-dev/main/datalog.go:587 +0xb4

goroutine 19 [chan receive]:
main.gpsAttitudeSender()
        /root/stratux-dev/main/gps.go:1904 +0xec
created by main.pollGPS
        /root/stratux-dev/main/gps.go:2017 +0x4c

goroutine 45 [chan receive]:
main.pollGPS()
        /root/stratux-dev/main/gps.go:2019 +0x6c
created by main.initGPS
        /root/stratux-dev/main/gps.go:2038 +0x10c

goroutine 46 [select]:
main.heartBeatSender()
        /root/stratux-dev/main/gen_gdl90.go:611 +0x334
created by main.main
        /root/stratux-dev/main/gen_gdl90.go:1438 +0x96c

goroutine 47 [IO wait, 117 minutes]:
net.runtime_pollWait(0x740c4e58, 0x72, 0x6536b0)
        /root/go/src/runtime/netpoll.go:160 +0x60
net.(*pollDesc).wait(0x109962f8, 0x72, 0x0, 0x0)
        /root/go/src/net/fd_poll_runtime.go:73 +0x34
net.(*pollDesc).waitRead(0x109962f8, 0x0, 0x0)
        /root/go/src/net/fd_poll_runtime.go:78 +0x30
net.(*netFD).accept(0x109962c0, 0x0, 0x652c60, 0x141a4870)
        /root/go/src/net/fd_unix.go:419 +0x21c
net.(*TCPListener).accept(0x1099c028, 0x2470d1ff, 0x0, 0x0)
        /root/go/src/net/tcpsock_posix.go:132 +0x30
net.(*TCPListener).AcceptTCP(0x1099c028, 0xae214, 0x0, 0x0)
        /root/go/src/net/tcpsock.go:209 +0x9c
net/http.tcpKeepAliveListener.Accept(0x1099c028, 0x0, 0x0, 0x0, 0x0)
        /root/go/src/net/http/server.go:2608 +0x3c
net/http.(*Server).Serve(0x10a58000, 0x654fa8, 0x1099c028, 0x0, 0x0)
        /root/go/src/net/http/server.go:2273 +0x200
net/http.(*Server).ListenAndServe(0x10a58000, 0x0, 0x0)
        /root/go/src/net/http/server.go:2219 +0xf4
net/http.ListenAndServe(0x4ca879, 0x3, 0x0, 0x0, 0x0, 0x0)
        /root/go/src/net/http/server.go:2351 +0x90
main.managementInterface()
        /root/stratux-dev/main/managementinterface.go:638 +0x454
created by main.main
        /root/stratux-dev/main/gen_gdl90.go:1440 +0x980

goroutine 20 [chan receive, 4186 minutes]:
main.(*uibroadcaster).writer(0x109f8080)
        /root/stratux-dev/main/uibroadcast.go:53 +0x48
created by main.NewUIBroadcaster
        /root/stratux-dev/main/uibroadcast.go:32 +0x13c

goroutine 21 [chan receive]:
main.(*uibroadcaster).writer(0x109f80a0)
        /root/stratux-dev/main/uibroadcast.go:53 +0x48
created by main.NewUIBroadcaster
        /root/stratux-dev/main/uibroadcast.go:32 +0x13c

goroutine 22 [chan receive]:
main.(*uibroadcaster).writer(0x109f80c0)
        /root/stratux-dev/main/uibroadcast.go:53 +0x48
created by main.NewUIBroadcaster
        /root/stratux-dev/main/uibroadcast.go:32 +0x13c

goroutine 23 [chan receive, 4186 minutes]:
main.(*uibroadcaster).writer(0x109f80e0)
        /root/stratux-dev/main/uibroadcast.go:53 +0x48
created by main.NewUIBroadcaster
        /root/stratux-dev/main/uibroadcast.go:32 +0x13c

goroutine 24 [chan receive]:
main.(*uibroadcaster).writer(0x109f8100)
        /root/stratux-dev/main/uibroadcast.go:53 +0x48
created by main.NewUIBroadcaster
        /root/stratux-dev/main/uibroadcast.go:32 +0x13c

goroutine 48 [chan receive]:
main.monitorDHCPLeases()
        /root/stratux-dev/main/network.go:460 +0x60
created by main.initNetwork
        /root/stratux-dev/main/network.go:643 +0x194

goroutine 50 [semacquire]:
sync.runtime_Semacquire(0x109dc904)
        /root/go/src/runtime/sema.go:47 +0x24
sync.(*Mutex).Lock(0x109dc900)
        /root/go/src/sync/mutex.go:85 +0x1fc
main.sleepMonitor()
        /root/stratux-dev/main/network.go:545 +0x4fc
created by main.initNetwork
        /root/stratux-dev/main/network.go:645 +0x1bc

goroutine 51 [chan receive]:
main.networkStatsCounter()
        /root/stratux-dev/main/network.go:563 +0x8c
created by main.initNetwork
        /root/stratux-dev/main/network.go:646 +0x1d0

goroutine 52 [select]:
main.serialOutWatcher()
        /root/stratux-dev/main/network.go:237 +0x870
created by main.initNetwork
        /root/stratux-dev/main/network.go:647 +0x1e4

goroutine 53 [chan receive]:
main.networkOutWatcher()
        /root/stratux-dev/main/network.go:224 +0x40
created by main.initNetwork
        /root/stratux-dev/main/network.go:648 +0x1f8

goroutine 54 [chan receive]:
main.printStats()
        /root/stratux-dev/main/gen_gdl90.go:1203 +0x9c
created by main.main
        /root/stratux-dev/main/gen_gdl90.go:1446 +0x998

goroutine 55 [chan receive]:
main.cpuTempMonitor()
        /root/stratux-dev/main/gen_gdl90.go:733 +0x4c
created by main.main
        /root/stratux-dev/main/gen_gdl90.go:1449 +0x9ac

goroutine 25 [chan receive]:
main.icmpEchoSender(0x10998170)
        /root/stratux-dev/main/network.go:469 +0x70
created by main.sleepMonitor
        /root/stratux-dev/main/network.go:506 +0xfc

goroutine 1418 [IO wait]:
net.runtime_pollWait(0x740c4930, 0x72, 0x11e8b000)
        /root/go/src/runtime/netpoll.go:160 +0x60
net.(*pollDesc).wait(0x11174cb8, 0x72, 0x0, 0x0)
        /root/go/src/net/fd_poll_runtime.go:73 +0x34
net.(*pollDesc).waitRead(0x11174cb8, 0x0, 0x0)
        /root/go/src/net/fd_poll_runtime.go:78 +0x30
net.(*netFD).Read(0x11174c80, 0x11e8b000, 0x1000, 0x1000, 0x0, 0x6536b0, 0x109100c0)
        /root/go/src/net/fd_unix.go:243 +0x1f0
net.(*conn).Read(0x1114aff8, 0x11e8b000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
        /root/go/src/net/net.go:173 +0xb8
net/http.(*connReader).Read(0x12f3b740, 0x11e8b000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
        /root/go/src/net/http/server.go:586 +0x1d8
bufio.(*Reader).fill(0x116f2360)
        /root/go/src/bufio/bufio.go:97 +0x1c4
bufio.(*Reader).ReadSlice(0x116f2360, 0xa, 0x0, 0x0, 0x0, 0x0, 0x0)
        /root/go/src/bufio/bufio.go:330 +0x264
bufio.(*Reader).ReadLine(0x116f2360, 0x0, 0x0, 0x0, 0x1307aa00, 0x0, 0x0)
        /root/go/src/bufio/bufio.go:359 +0x60
net/textproto.(*Reader).readLineSlice(0x10a0e260, 0x0, 0x0, 0x0, 0x0, 0x0)
        /root/go/src/net/textproto/reader.go:55 +0x58
net/textproto.(*Reader).ReadLine(0x10a0e260, 0x0, 0x0, 0x0, 0x0)
        /root/go/src/net/textproto/reader.go:36 +0x34
net/http.readRequest(0x116f2360, 0x400, 0x11481f80, 0x0, 0x0)
        /root/go/src/net/http/request.go:793 +0x80
net/http.(*conn).readRequest(0x11174d40, 0x6553a0, 0x12f3b780, 0x0, 0x0, 0x0)
        /root/go/src/net/http/server.go:765 +0x3d4
net/http.(*conn).serve(0x11174d40, 0x6553a0, 0x12f3b780)
        /root/go/src/net/http/server.go:1532 +0x910
created by net/http.(*Server).Serve
        /root/go/src/net/http/server.go:2293 +0x470

goroutine 67 [syscall]:
syscall.Syscall(0x3, 0x13, 0x10bb1758, 0x8a8, 0x6f, 0x0, 0x0)
        /root/go/src/syscall/asm_linux_arm.s:17 +0x8
syscall.read(0x13, 0x10bb1758, 0x8a8, 0x8a8, 0x6f, 0x0, 0x0)
        /root/go/src/syscall/zsyscall_linux_arm.go:783 +0x78
syscall.Read(0x13, 0x10bb1758, 0x8a8, 0x8a8, 0x6f, 0x0, 0x0)
        /root/go/src/syscall/syscall_unix.go:161 +0x4c
os.(*File).read(0x1099ca90, 0x10bb1758, 0x8a8, 0x8a8, 0x6f, 0x0, 0x0)
        /root/go/src/os/file_unix.go:228 +0x54
os.(*File).Read(0x1099ca90, 0x10bb1758, 0x8a8, 0x8a8, 0x460ba400, 0x0, 0x0)
        /root/go/src/os/file.go:101 +0x7c
github.com/tarm/serial.(*Port).Read(0x1099ca98, 0x10bb1758, 0x8a8, 0x8a8, 0x40accccd, 0x0, 0x0)
        /root/go_path/src/github.com/tarm/serial/serial_posix.go:180 +0x50
bufio.(*Scanner).Scan(0x10a8ff84, 0x6d)
        /root/go/src/bufio/scan.go:208 +0x3b4
main.gpsSerialReader()
        /root/stratux-dev/main/gps.go:1780 +0xe0
created by main.pollGPS
        /root/stratux-dev/main/gps.go:2024 +0xd0

goroutine 1401 [chan receive]:
main.handleStatusWS(0x10a58370)
        /root/stratux-dev/main/managementinterface.go:170 +0xb4
golang.org/x/net/websocket.Server.serveWebSocket(0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /root/go_path/src/golang.org/x/net/websocket/server.go:89 +0x274
golang.org/x/net/websocket.Server.ServeHTTP(0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /root/go_path/src/golang.org/x/net/websocket/server.go:70 +0x38
main.managementInterface.func2(0x654ea8, 0x10c517a0, 0x1155e000)
        /root/stratux-dev/main/managementinterface.go:594 +0x4c
net/http.HandlerFunc.ServeHTTP(0x50e7bc, 0x654ea8, 0x10c517a0, 0x1155e000)
        /root/go/src/net/http/server.go:1726 +0x34
net/http.(*ServeMux).ServeHTTP(0x6695a8, 0x654ea8, 0x10c517a0, 0x1155e000)
        /root/go/src/net/http/server.go:2022 +0x164
net/http.serverHandler.ServeHTTP(0x10a58000, 0x654ea8, 0x10c517a0, 0x1155e000)
        /root/go/src/net/http/server.go:2202 +0x150
net/http.(*conn).serve(0x11186640, 0x6553a0, 0x14244e00)
        /root/go/src/net/http/server.go:1579 +0xde8
created by net/http.(*Server).Serve
        /root/go/src/net/http/server.go:2293 +0x470

goroutine 1432 [chan receive]:
main.handleStatusWS(0x109ea550)
        /root/stratux-dev/main/managementinterface.go:170 +0xb4
golang.org/x/net/websocket.Server.serveWebSocket(0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /root/go_path/src/golang.org/x/net/websocket/server.go:89 +0x274
golang.org/x/net/websocket.Server.ServeHTTP(0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /root/go_path/src/golang.org/x/net/websocket/server.go:70 +0x38
main.managementInterface.func2(0x654ea8, 0x131b0f30, 0x10aebc80)
        /root/stratux-dev/main/managementinterface.go:594 +0x4c
net/http.HandlerFunc.ServeHTTP(0x50e7bc, 0x654ea8, 0x131b0f30, 0x10aebc80)
        /root/go/src/net/http/server.go:1726 +0x34
net/http.(*ServeMux).ServeHTTP(0x6695a8, 0x654ea8, 0x131b0f30, 0x10aebc80)
        /root/go/src/net/http/server.go:2022 +0x164
net/http.serverHandler.ServeHTTP(0x10a58000, 0x654ea8, 0x131b0f30, 0x10aebc80)
        /root/go/src/net/http/server.go:2202 +0x150
net/http.(*conn).serve(0x10fdc9c0, 0x6553a0, 0x10c143a0)
        /root/go/src/net/http/server.go:1579 +0xde8
created by net/http.(*Server).Serve
        /root/go/src/net/http/server.go:2293 +0x470

goroutine 1402 [chan receive]:
main.handleStatusWS(0x10c30370)
        /root/stratux-dev/main/managementinterface.go:170 +0xb4
golang.org/x/net/websocket.Server.serveWebSocket(0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /root/go_path/src/golang.org/x/net/websocket/server.go:89 +0x274
golang.org/x/net/websocket.Server.ServeHTTP(0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /root/go_path/src/golang.org/x/net/websocket/server.go:70 +0x38
main.managementInterface.func2(0x654ea8, 0x141d73b0, 0x10965300)
        /root/stratux-dev/main/managementinterface.go:594 +0x4c
net/http.HandlerFunc.ServeHTTP(0x50e7bc, 0x654ea8, 0x141d73b0, 0x10965300)
        /root/go/src/net/http/server.go:1726 +0x34
net/http.(*ServeMux).ServeHTTP(0x6695a8, 0x654ea8, 0x141d73b0, 0x10965300)
        /root/go/src/net/http/server.go:2022 +0x164
net/http.serverHandler.ServeHTTP(0x10a58000, 0x654ea8, 0x141d73b0, 0x10965300)
        /root/go/src/net/http/server.go:2202 +0x150
net/http.(*conn).serve(0x110e9640, 0x6553a0, 0x12f3af20)
        /root/go/src/net/http/server.go:1579 +0xde8
created by net/http.(*Server).Serve
        /root/go/src/net/http/server.go:2293 +0x470

goroutine 1361 [IO wait]:
net.runtime_pollWait(0x740c45e8, 0x72, 0x1424e000)
        /root/go/src/runtime/netpoll.go:160 +0x60
net.(*pollDesc).wait(0x11187b38, 0x72, 0x0, 0x0)
        /root/go/src/net/fd_poll_runtime.go:73 +0x34
net.(*pollDesc).waitRead(0x11187b38, 0x0, 0x0)
        /root/go/src/net/fd_poll_runtime.go:78 +0x30
net.(*netFD).Read(0x11187b00, 0x1424e000, 0x1000, 0x1000, 0x0, 0x6536b0, 0x109100c0)
        /root/go/src/net/fd_unix.go:243 +0x1f0
net.(*conn).Read(0x109e5418, 0x1424e000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
        /root/go/src/net/net.go:173 +0xb8
net/http.(*connReader).Read(0x1333e600, 0x1424e000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
        /root/go/src/net/http/server.go:586 +0x1d8
bufio.(*Reader).fill(0x10bdce40)
        /root/go/src/bufio/bufio.go:97 +0x1c4
bufio.(*Reader).ReadSlice(0x10bdce40, 0xa, 0x0, 0x0, 0x0, 0x0, 0x0)
        /root/go/src/bufio/bufio.go:330 +0x264
bufio.(*Reader).ReadLine(0x10bdce40, 0x0, 0x0, 0x0, 0x64fc00, 0x0, 0x0)
        /root/go/src/bufio/bufio.go:359 +0x60
net/textproto.(*Reader).readLineSlice(0x10a98020, 0x0, 0x0, 0x0, 0x0, 0x0)
        /root/go/src/net/textproto/reader.go:55 +0x58
net/textproto.(*Reader).ReadLine(0x10a98020, 0x0, 0x0, 0x0, 0x0)
        /root/go/src/net/textproto/reader.go:36 +0x34
net/http.readRequest(0x10bdce40, 0x0, 0x11e99200, 0x0, 0x0)
        /root/go/src/net/http/request.go:793 +0x80
net/http.(*conn).readRequest(0x11187b80, 0x6553a0, 0x1333e640, 0x0, 0x0, 0x0)
        /root/go/src/net/http/server.go:765 +0x3d4
net/http.(*conn).serve(0x11187b80, 0x6553a0, 0x1333e640)
        /root/go/src/net/http/server.go:1532 +0x910
created by net/http.(*Server).Serve
        /root/go/src/net/http/server.go:2293 +0x470

goroutine 1311 [runnable, locked to thread]:
github.com/jpoirier/gortlsdr._Cfunc_rtlsdr_read_sync(0x73f016e8, 0x116a6000, 0x40000, 0x110464bc, 0x0)
        github.com/jpoirier/gortlsdr/_obj/_cgo_gotypes.go:382 +0x3c
github.com/jpoirier/gortlsdr.(*Context).ReadSync(0x73f016e8, 0x116a6000, 0x40000, 0x40000, 0x40000, 0x40000, 0x0, 0x0)
        /root/go_path/src/github.com/jpoirier/gortlsdr/rtlsdr.go:550 +0x18c
main.(*UAT).read(0x10c140a0)
        /root/stratux-dev/main/sdr.go:148 +0x200
created by main.createUATDev
        /root/stratux-dev/main/sdr.go:404 +0x228
```

**Comment by cyoung** [comment]
Good find.

**Comment by westphae** [comment]
I've been encountering this panic also.  It occurs regularly when gen_gdl90 is built with Go versions 1.6 or higher.  Runtime detection of concurrent map reads/writes was introduced in Go 1.6:  see https://golang.org/doc/go1.6#runtime.  @AvSquirrel , were you compiling with a more current version of Go?

pingResponse isn't the only map that suffers this panic.  I protected pingResponse with a mutex in my fork and subsequently also had a panic due to concurrent read/write of outSockets at https://github.com/cyoung/stratux/blob/master/main/network.go#L472.  My fork implements AHRS and runs at a higher CPU load, which I believe tends to make these panics occur more frequently.

**Comment by Ergonomicmike** [comment]
Regarding CPU load under AHRS: Will it be necessary to change the governor for AHRS? Last I knew, the Stratux was throttled back quite a bit, but loafing along with normal ADS-B loads. (Hmmm... Maybe this is why sqlite logging lags?)

**Comment by westphae** [comment]
I don't think it will be necessary to increase clock speed for AHRS.  We want the battery to last as long as possible!  I have the algorithm set to update more often than necessary while testing, so it's at a higher load now than it really needs to be.

But I suspect that the higher load increases the rate of thread collisions for this issue.  I had tried to reproduce this panic a few weeks ago without AHRS (so cyoung/stratux code but built with Go 1.7.5) and didn't get a panic after running for 24+ hours.  With my fork (no significant changes to network.go) it can happen within an hour without the mutex, even if the load is only 1.7 or so.

Mutex protection seems to resolve this issue for the pingResponse map, but before submitting a pull request I wanted to protect outSockets also.

**Comment by westphae** [comment]
Looking through network.go, netMutex already exists to protect outSockets.  netMutex seems mainly intended to protect outSockets from simultaneous writes (which also happens to protect some of the reads).  In two places where outSockets is only read from, it is not protected by netMutex, namely https://github.com/cyoung/stratux/blob/master/main/network.go#L291 and https://github.com/cyoung/stratux/blob/master/main/network.go#L472.  The latter is the one that usually gives me grief).

How about if we change netMutex to a sync.RWMutex and protect the other reads with RLock()?  pingResponse also needs a sync.RWMutex.  Potentially it could be covered by netMutex--it is only written to at https://github.com/cyoung/stratux/blob/master/main/network.go#L524--but possibly it would lock for too long?  I set up a separate Mutex and in my limited test case that seems to do the trick so far.

**Comment by cyoung** [comment]
@westphae - just skimmed your branch, on increasing load causing collisions, I don't think it relates to system/CPU load so much as it does an increased packet rate (and more frequent calls to `sendToAllConnectedClients()`. Increased number of clients connected, packet rate, and decreasing the traffic packet maximum size (https://github.com/cyoung/stratux/commit/3e663c89c8a812cb5f3f482f0c933bc46152c842) were all probably contributing factors.

**Comment by cyoung** [comment]
Since most of the time that `netMutex` should be locked is within `messageQueueSender()`, and all of the other events which lock it should not occur often - using `netMutex` to protect all three of the network maps seems like the simplest approach. https://github.com/cyoung/stratux/commit/0360e506ec3336be77006c5ec3c4d2b3b3a860fc

* `icmpEchoSender()`
* `sleepMonitor()`
* `getNetworkStats()`
* `handleClientsGetRequest()`

This isn't the ideal approach, since it allows the network output stream to be affected by external forces (ICMP packets, HTTP requests) - so it would be easy to construct a DoS attack on the system.
Looking at the code again, it seems we had this problem all along. So this doesn't fix the DoS vulnerability issue, but should fix the normal operations issue. Let's try this approach for now and revisit the network code some time soon in the future, since there is a lot that can be simplified here.

**Comment by cyoung** [comment]
http://updates.stratux.me/builds/update-stratux-v1.2r1-0360e506ec.sh

**Comment by AvSquirrel** [comment]
@cyoung - I'll merge 0360e50 into my TIS-B suppression branch and give it a shot. (Haven't had any time to look at it since my last commit, but maybe I'll poke around this week.) Thanks!


@westphae - I had been compiling with go1.7.2 linux/arm, and panics were repeatable, taking anywhere from two hours to three days to bring down gen_gdl90. However, I see in the [go release notes](https://golang.org/doc/devel/release.html) that:

>go1.7.2 should not be used. It was tagged but not fully released. The release was deferred due to a last minute bug report.

and that [go1.8](https://golang.org/doc/go1.8) released a couple weeks ago, with an improved compiler and a more responsive garbage collector (i.e. fewer / shorter pauses). I'm going to install and give it a whirl.

**Comment by AvSquirrel** [comment]
Huh. That was unexpected.

On go1.8 on 0ee94ff92 (old network code) panicked after 24 seconds, but on a different map variable ([`outSockets`](https://github.com/cyoung/stratux/blob/6198fc2ce34f8b230066d42d21672784ef19cd8d/main/network.go#L472)).

```
fatal error: concurrent map iteration and map write

goroutine 74 [running]:
runtime.throw(0x467ae5, 0x26)
        /root/go/src/runtime/panic.go:596 +0x70 fp=0x10a14630 sp=0x10a14624
runtime.mapiternext(0x10a14760)
        /root/go/src/runtime/hashmap.go:737 +0x6cc fp=0x10a1467c sp=0x10a14630
main.icmpEchoSender(0x10a6c0a0)
        /root/stratux-dev/main/network.go:472 +0x144 fp=0x10a147e4 sp=0x10a1467c
runtime.goexit()
        /root/go/src/runtime/asm_arm.s:1017 +0x4 fp=0x10a147e4 sp=0x10a147e4
created by main.sleepMonitor
        /root/stratux-dev/main/network.go:506 +0x68
```

I think I'll revert to go1.7.5 for now.

**Comment by cyoung** [comment]
"concurrent map iteration and map write", that's actually another issue that needs to be fixed.

**Comment by AvSquirrel** [comment]
@cyoung, FWIW I reinstalled go1.8 and compiled with the mutex-protected network code from 0360e50. 

Ran overnight (10 hours) with no obvious issues. I'll continue bench testing and hope to take if flying this weekend.

**Comment by westphae** [comment]
@AvSquirrel I had the panic on outSockets also.  I don't remember having a panic on simultaneous iteration and write, but I guess that's just a variation on reading and writing.  I've had the same results whether using 1.7.5 or 1.8.

@cyoung I thought maybe the increased load caused everything to slow down, increasing the likelihood of a collision, but I could be wrong.  I had the panics even with only one client connected, and I don't think the AHRS is sending any more data.  In any case, protecting all the reads/writes with a mutex solved it for me.  I ran for a few days straight without any more panics, as well as over a few dozens of hours separately on quite a few test flights and bench runs.  I only protected outSockets and pingResponse, and decided that one mutex per map (i.e. just one additional mutex, though I made it a RWMutex instead of a plain Mutex) wasn't overly complicated and should minimize unnecessary blocking.  So far it's been trouble-free.  You said there's a third map you protected?  If your solution works, I'll change to your approach next time I merge your code into mine just to minimize differences.

**Comment by westphae** [comment]
Ah, I see it: dhcpLeases is the third map.  I guess that one can't be written to and read from simultaneously anyway.

**Comment by cyoung** [comment]
That's it - was not necessary but it might be read from in web calls in the future.

**Comment by cyoung** [comment]
Testing seems good, changed as above and released in v1.2r2.
